database/migrate.go

```diff
@@ -0,0 +1,202 @@
+package database
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+
+	"github.com/golang-migrate/migrate/v4"
+	// required by migrate.New... to parse migration files directory
+	_ "github.com/golang-migrate/migrate/v4/source/file"
+
+	migratepg "github.com/golang-migrate/migrate/v4/database/postgres"
+)
+
+// supported migrate operations
+const (
+	defaultFilesDir = "dbmigrations"
+	operationUp     = "up"
+	operationDown   = "down"
+	operationForce  = "force"
+)
+
+type operationFn func(*MigrationRunner, OperationData) error
+
+var supportedOperations = map[string]operationFn{
+	operationUp:    runUp,
+	operationDown:  runDown,
+	operationForce: runForce,
+}
+
+// OperationData contains information about the migration operation to be performed.
+type OperationData struct {
+	ID           string
+	ForceVersion int
+}
+
+// MigrationRunner is responsible for managing and running database migrations.
+type MigrationRunner struct {
+	mgr      *migrate.Migrate
+	filesDir string
+	logger   logger
+}
+
+// Option represents a function that configures a MigrationRunner.
+type Option func(runner *MigrationRunner)
+
+// WithLogger sets a custom logger for the MigrationRunner.
+// If not provided, a noopLogger will be used by default.
+func WithLogger(logger logger) Option {
+	return func(runner *MigrationRunner) {
+		runner.logger = logger
+	}
+}
+
+// WithFilesDir sets a custom directory containing migration files for the MigrationRunner.
+// If not provided, the default directory "dbmigrations" will be used.
+func WithFilesDir(filesDir string) Option {
+	return func(runner *MigrationRunner) {
+		runner.filesDir = filesDir
+	}
+}
+
+// NewMigrationRunner creates a new MigrationRunner with the given database connection string (dsn) and options.
+func NewMigrationRunner(dsn string, opts ...Option) (*MigrationRunner, error) {
+	db, err := sql.Open("postgres", dsn)
+	if err != nil {
+		return nil, fmt.Errorf("sql open dsn: %v", err)
+	}
+
+	driver, err := migratepg.WithInstance(db, &migratepg.Config{})
+	if err != nil {
+		return nil, fmt.Errorf("building postgres driver for db migrations: %w", err)
+	}
+
+	runner := &MigrationRunner{}
+
+	for _, opt := range opts {
+		opt(runner)
+	}
+
+	if runner.filesDir == "" {
+		runner.filesDir = defaultFilesDir
+	}
+
+	if runner.logger == nil {
+		runner.logger = &noopLogger{}
+	}
+
+	mgr, err := migrate.NewWithDatabaseInstance(
```

Comment:
we can simplify this a bit, and use `migrate.New("file://...", "connection string")` instead of initializing db connection with sql.Open and then passing it no NewWithDatabaseInstance.